docs/development/testguide.rst

```diff
@@ -117,6 +100,27 @@ This can be used in conjunction with the ``-P`` option provided by the
 `pytest-filter-subpackage <https://github.com/astropy/pytest-filter-subpackage>`_
 plugin to run just part of the test suite.
 
+Using tox to set up a development environment
+---------------------------------------------
+
+As noted, running ``tox`` can be quite slow, yet it has the advantage over the
+other methods that it puts together a consistent set of dependencies.  One can
+speed up matters by using ``tox`` to set up a virtual environment for
+development and then use direct ``pytest`` commands within that environment.
+For instance::
+
+    tox -e test-alldeps --develop --notest
```

Comment:
Is this really something that we should be recommending? It seems like a very weird way to spin up a virtualenv and people might end up here in all manner of weird ways and not really understand what they are doing, (i.e. virtualenv inside conda etc).

The justification from the sentence below is:

> Now, one can run any of the ``pytest`` commands below to only run tests that are relevant to the development one is doing.

You can already customise the pytest behaviour by passing extra args to the tox command, and you can skip the (slow) sdist creation and build with the `--develop` flag. I really think this should at least come with a large health warning on the top. I see the utility if you really have to debug a failure in one specific tox env and you really want to get in there with a python interpreter, but I really struggle to see the justification for doing this to run anything but the most spectacularly complex pytest commands.

Reply:
@Cadair - this came from real life problems from experienced developers (@mhvk and myself) having trouble just being able to run tests in an agile manner. What we need is a way to be able to have a one-time setup that reproduces accurately the travis test environment and then be able to re-run tests (over and over again, sometimes) in a few seconds, ala:
```
pytest astropy/time --pdb -k test_darn_its_still_failing
```
In discussions referenced in #10303, this was the best we came up with.  If you have a better suggestion, please do tell!!  That would be great.  You can see already below that I also am not quite happy with the conda/virtualenv mixture that is likely coming out of this.

Reply:
And note there are still outstanding problems with `pytest` as far as I can tell. If I just blindly install `pytest`, that gives me 5.4.0 and the dreaded error in test collection:
```
E   pytest.PytestDeprecationWarning: direct construction of DocTestModulePlus has been deprecated, please use DocTestModulePlus.from_parent
```
Maybe I'm doing something wrong, but if I am having problems that means others may be in the same boat.

Reply:
@taldcroft - what version of pytest-doctestplus are you using? The issues should be fixed with v0.7.0

Reply:
I was using 0.5.  On conda (astropy / conda-forge) the latest available seems to be 0.6.1.

Reply:
the "correct" solution here would be to use `tox ... --develop` but there seems to be a bug with that and isolated builds. I am investigating.

Reply:
Thanks!  As @taldcroft noted, what would be most helpful is to regain the option to have fast execution of a single test (or test file). There are also advantages to using `pytest` directly, e.g., if I start typing `pytest --` and then press tab, I get a nice display of all options because `bash` knows where to look.

p.s. Like @taldcroft, I actually prefer to just rely on what is installed on my machine already, and I'd gladly go back to just having a single development enviroment created with `--system-site-packages` and doing `pip install -e .`. Since Debian is pretty good at having a nice complement, for me the tox route is mostly for those times when pytest and doctestplus get out of sync, etc.  But here the goal is to try to describe a method that is fast and works reliable in as many situations as possible...

Reply:
I have opened an issue against tox for the bug with `--develop` https://github.com/tox-dev/tox/issues/1595

As far as I am concerned that is the best way to achieve "fast and reliable" for a testing environment. It obviously should be possible to just run a `pytest` in the repo, and we need to do as much as we can to make that easy (i.e. `pip install -e .[all,test]; pytest` should always work), however I really think that as far as documentation to make it easy for people to contribute we should be heavily encouraging using tox.

Reply:
@Cadair - yes, if there is a good way just with `tox`, that would be preferable! The one advantage of the present suggestion is that it shows, indirectly, a bit how `tox` actually works.

Reply:
The suggested workaround on the tox issue is to put the build deps from `pyproject.toml` into the `deps=` list in `tox.ini` which does seem to work, although isn't an ideal solution imo.

Reply:
There is a tox rewrite underway, I just asked upstream if the bug will be properly addressed in the new version 🤞🏽

Reply:
@Cadair, @astrojuanlu - this PR got stuck because my recommendation for fast testing with `tox --develop --no-test` and then enabling the virtual environment was clearly not ideal, but on the other hand there didn't seem to be a good alternative. I see that @Cadair's tox issue https://github.com/tox-dev/tox/issues/1595 has been closed because this is fixed in "the rewrite branch", though I think that implies tox 4, which is not out yet.  It certainly still fails with ~current tox.

Since it now has been more than a year with no other effective solution, my sense is to get this PR in and revisit when there is a better solution. But I'm also happy to just let it auto-close - after all, my own work flow is now fine...

Reply:
My feeling is that tox 4 is "almost there", but that can still mean a few months. This is a net improvement on the current docs, so I'm 👍🏽  on getting it in.